tfhe/src/shortint/server_key/tests/noise_level.rs

```diff
@@ -0,0 +1,263 @@
+        },
+        &|_, _| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_1_ct_1_scalar_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext, scalar: u8) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &Ciphertext, u8) -> Ciphertext,
+                   g: &dyn Fn(NoiseLevel, u8) -> NoiseLevel| {
+        assert!(f(sk, ct, scalar).noise_level == g(ct.noise_level, scalar));
+    };
+
+    test_fn(&ServerKey::unchecked_scalar_add, &|a, _| a);
+    test_fn(&ServerKey::unchecked_scalar_sub, &|a, _| a);
+    test_fn(&ServerKey::unchecked_scalar_mul, &|a, b| a * b as usize);
+    if scalar != 0 {
+        test_fn(&ServerKey::unchecked_scalar_div, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+        test_fn(&ServerKey::unchecked_scalar_mod, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_bitand, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitor, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitxor, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    if scalar < 8 {
+        test_fn(&ServerKey::unchecked_scalar_left_shift, &|a, b| {
+            a * (1 << b as usize)
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_right_shift, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+}
+
+fn test_1_ct_1_scalar_assign_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext, scalar: u8) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &mut Ciphertext, u8),
+                   g: &dyn Fn(NoiseLevel, u8) -> NoiseLevel| {
+        let mut clone = ct.clone();
+        f(sk, &mut clone, scalar);
+        assert!(clone.noise_level == g(ct.noise_level, scalar));
+    };
+
+    test_fn(&ServerKey::unchecked_scalar_add_assign, &|a, _| a);
+    test_fn(&ServerKey::unchecked_scalar_sub_assign, &|a, _| a);
+    test_fn(&ServerKey::unchecked_scalar_mul_assign, &|a, b| {
+        a * b as usize
+    });
+    if scalar != 0 {
+        test_fn(&ServerKey::unchecked_scalar_div_assign, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+        test_fn(&ServerKey::unchecked_scalar_mod_assign, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_bitand_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitor_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitxor_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    if scalar < 8 {
+        test_fn(&ServerKey::unchecked_scalar_left_shift_assign, &|a, b| {
+            a * (1 << b as usize)
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_right_shift_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+}
+
+#[test]
+fn test_noise_level_propagation() {
+    let keys = KEY_CACHE.get_from_param(PARAM_MESSAGE_2_CARRY_2_KS_PBS);
+    let (ck, sk) = (keys.client_key(), keys.server_key());
+
```

Comment:
let's have several test runs with the randomized values in addition to the 0s

Reply:
Done